src/Servers/Kestrel/Core/src/Internal/Http/PathNormalizer.cs

```diff
@@ -50,193 +52,144 @@ public static string DecodePath(Span<byte> path, bool pathEncoded, string rawTar
-                        *src = ByteSlash;
-
-                        if (dst > start)
-                        {
-                            do
-                            {
-                                dst--;
-                            } while (dst > start && *dst != ByteSlash);
-                        }
-
-                        continue;
-                    }
-                    else if (ch2 == ByteDot && ch3 == ByteSlash)
-                    {
-                        // B.  if the input buffer begins with a prefix of "/./" or "/.",
-                        //     where "." is a complete path segment, then replace that
-                        //     prefix with "/" in the input buffer; otherwise,
-                        src += 2;
-                        continue;
-                    }
-
-                    break;
-                default:
-                    ch2 = *(src + 1);
-                    ch3 = *(src + 2);
-                    ch4 = *(src + 3);
+                return src.Length;
+            }
+            if (nextDotSegmentIndex < 0)
+            {
+                // Copy the remianing src to dst, and return.
+                src[readPointer..].CopyTo(src[writtenLength..]);
+                writtenLength += src.Length - readPointer;
+                return writtenLength;
+            }
+            else if (nextDotSegmentIndex > 0)
+            {
+                // Copy until the next segment excluding the trailer.
+                Unsafe.CopyBlock(ref src[writtenLength], ref src[readPointer], (uint)nextDotSegmentIndex);
+                writtenLength += nextDotSegmentIndex;
+                readPointer += nextDotSegmentIndex;
+            }
 
-                    if (ch2 == ByteDot && ch3 == ByteDot && ch4 == ByteSlash)
+            // Case of /../ or /./
+            var remainingLength = src.Length - readPointer;
+            if (remainingLength > 3)
+            {
+                var nextIndex = readPointer + 2;
+                if (src[nextIndex] == ByteSlash)
+                {
+                    readPointer = nextIndex;
+                }
+                else if (MemoryMarshal.CreateSpan(ref src[nextIndex], 2).StartsWith(dotSlash))
+                {
+                    // Remove the last segment and replace the path with /
+                    var lastIndex = MemoryMarshal.CreateSpan(ref src[0], writtenLength).LastIndexOf(ByteSlash);
+
+                    // Move write pointer to the end of the previous segment without / or to start position
+                    writtenLength = int.Max(0, lastIndex);
+
+                    // Move the read pointer to the next segments beginning including /
+                    readPointer += 3;
+                }
+                else
+                {
+                    // Not a dot segment, copy the matched /. and bump the read pointer
+                    slashDot.CopyTo(src[writtenLength..]);
+                    writtenLength += 2;
```

Comment:
Can we technically copy 3 values since we know the 3rd character isn't a `.` or `/`?

Reply:
I think yes, though it also requires then to increment the read pointer.

Reply:
Sure, but that's fine?
In the case below where we have `else if (remainingLength == 3)` then we can avoid an extra iteration of the loop.

Reply:
That is a very good point for the 3 chars, let me update it.

Reply:
Updated as suggested and updated the performance measurements. Please feel free to have a look, it looks reasonable to me.